Document the k6/experimental/tracing module (#1020)
grafana/k6-docs: src/data/markdown/docs/02 javascript api/07 k6-experimental/04 tracing/02 Client.md

---
title: 'Client'
excerpt: 'Client is a HTTP client attaching tracing information to its requests.'
---

> `Client` is an HTTP client class that attaches tracing information to its requests. It lets users include a tracing context in their HTTP requests so that tracing backends (such as Grafana Tempo) can incorporate their results.

Reviewer: ```suggestion
`Client` is an HTTP client class that attaches tracing information to its requests. It lets users include a tracing context in their HTTP requests so that tracing backends (such as [Grafana Tempo](https://grafana.com/oss/tempo/)) can incorporate their results.
```

 // the `instrumentHTTP` function, as it leaves the
 // imported standard http module untouched. Thus,
 // one can still perform non-instrumented HTTP calls
 // using it.
 res = http.post("http://httpbin.org/post", JSON.stringify(testData), {
 headers: { "X-Example-Header": "noninstrumented/post" },
 });
 check(res, {
 "status is 200": (r) => r.status === 200,
 });

 res = instrumentedHTTP.del("http://httpbin.org/delete", null, {
 headers: { "X-Example-Header": "instrumented/delete" },
 });
 check(res, {
 "status is 200": (r) => r.status === 200,
 });
};
```

</CodeGroup>

## HTTP module functions equivalents

The `Client` class exposes the same API as the standard `http` module. Except for the `batch` method, which is absent from `Client`. The following table lists the `Client` methods which have an equivalent in the standard `http` module:

| Method | HTTP equivalent | Description |
| :---------------------------------------------- | :------------------------------------------------ | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `Client.del(url, [body], [params])` | [http.del](/javascript-api/k6-http/del)** | Performs an instrumented HTTP DELETE request. The method has the same prototype as the `http.del` function and should transparently act as a drop-in replacement for it. |

Reviewer: What is the meaning of the `**` after the `http` methods? I was expecting some footnote, but didn't see any.

Author: Typo inherited from copy pasting 🤦🏻 Fixed 👍🏻

---
title: 'Client'
excerpt: 'Client is a HTTP client attaching tracing information to its requests.'
---

> `Client` is an HTTP client class that attaches tracing information to its requests. Use it to include a tracing context in HTTP requests so that tracing backends (such as [Grafana Tempo](https://grafana.com/oss/tempo/)) can incorporate their results.

Reviewer 2: more nitpicking: I would prefer to (I think I did have this discussion with someone else) not use the word *class* as it has specific meanings for people coming from non JS and in js it has specific meanings and arguably we do not have classes *at all* within k6, for now. 

There isn't really a go API to create classes - we only have pure constructors which is what kt/experimental/tracing uses.

So I would recommend using `constructor`.

Author: Really good point indeed 👍🏻

> `Client` is an HTTP client class that attaches tracing information to its requests. It lets users include a tracing context in their HTTP requests so that tracing backends (such as [Grafana Tempo](https://grafana.com/oss/tempo/)) can incorporate their results.

Reviewer 3: ```suggestion
`Client` is an HTTP client class that attaches tracing information to its requests. Use it to include a tracing context in HTTP requests so that tracing backends (such as [Grafana Tempo](https://grafana.com/oss/tempo/)) can incorporate their results.
```

>  The `Client` class acts as a drop-in replacement for the standard `http` module and will attach a [trace context](https://www.w3.org/TR/trace-context/) to the requests headers, and add a `trace_id` to HTTP-related k6 output's data points metadata. It currently supports the [W3C Trace Context](https://www.w3.org/TR/trace-context/) and [Jaeger](https://www.jaegertracing.io/docs/1.21/client-libraries/#propagation-format) trace context propagation formats. 

Reviewer 3: ```suggestion
 The `Client` class acts as a drop-in replacement for the standard `http` module and attaches a [trace context](https://www.w3.org/TR/trace-context/) to the requests headers, and add a `trace_id` to HTTP-related k6 output's data points metadata. It currently supports the [W3C Trace Context](https://www.w3.org/TR/trace-context/) and [Jaeger](https://www.jaegertracing.io/docs/1.21/client-libraries/#propagation-format) trace context propagation formats. 
```

The `Client` constructor accepts an [`Options`](/javascript-api/k6-experimental/tracing/options) object as its only parameter.

## Example

> This example demonstrates how to instantiate a tracing client and use it to instrument HTTP calls with trace context headers. It also illustrates how the standard `http` module can still be used side-by-side with it to perform non-instrumented HTTP calls.

Reviewer 3: ```suggestion
This example demonstrates how to instantiate a tracing client and use it to instrument HTTP calls with trace context headers. It also illustrates how you can use it alongside the standard `http` module to perform non-instrumented HTTP calls.
```

| `Client.patch(url, [body], [params])` | [http.patch](/javascript-api/k6-http/patch) | Performs an instrumented HTTP PATCH request. The method has the same prototype as the `http.patch` function and should transparently act as a drop-in replacement for it. |
| `Client.post(url, [body], [params])` | [http.post](/javascript-api/k6-http/post) | Performs an instrumented HTTP POST request. The method has the same prototype as the `http.post` function and should transparently act as a drop-in replacement for it. |
| `Client.put(url, [body], [params])` | [http.put](/javascript-api/k6-http/head) | Performs an instrumented HTTP HEAD request. The method has the same prototype as the `http.put` function and should transparently act as a drop-in replacement for it. |
| `Client.request(method, url, [body], [params])` | [http.request](/javascript-api/k6-http/request) | Performs an instrumented HTTP request. The method has the same prototype as the `http.request` function and should transparently act as a drop-in replacement for it. |

> | `Client.asyncRequest(method, url, [body], [params])` | [http.asyncRequest](/javascript-api/k6-http/asyncRequest) | Performs an instrumented HTTP asynchronous request. The method has the same prototype as the `http.asyncRequest` function and should transparently act as a drop-in replacement for it. |

Reviewer 2: ```suggestion
| `Client.asyncRequest(method, url, [body], [params])` | [http.asyncRequest](/javascript-api/k6-http/asyncrequest) | Performs an instrumented HTTP asynchronous request. The method has the same prototype as the `http.asyncRequest` function and should transparently act as a drop-in replacement for it. |
```

Paths are always lower case ... it seems

Author: My bad 🤦🏻 cheers!
